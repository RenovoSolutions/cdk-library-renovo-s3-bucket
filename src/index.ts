import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';

export interface IRenovoS3BucketProps {
  /**
   * Rules that define how Amazon S3 manages objects during their lifetime.
   */
  lifecycleRules: s3.LifecycleRule[];
  /**
   * The name of the bucket.
   */
  name?: string;
}

export class RenovoS3Bucket extends cdk.Construct {

  public readonly bucket: s3.Bucket;

  constructor(scope: cdk.Construct, id: string, props: IRenovoS3BucketProps) {
    super(scope, id);

    this.bucket = new s3.Bucket(this, 'bucket', {
      bucketName: props.name,
      // data retention
      objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
      versioned: true,
      lifecycleRules: props.lifecycleRules,
      // security
      encryption: s3.BucketEncryption.S3_MANAGED,
      enforceSSL: true,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
    });
  }
}
