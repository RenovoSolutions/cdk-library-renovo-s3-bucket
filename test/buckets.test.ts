import { SynthUtils } from '@aws-cdk/assert';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import { RenovoS3Bucket } from '../src/index';

test('Snapshot', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack', {
    env: {
      account: '123456789012', // not a real account
      region: 'us-east-1',
    },
  });

  new RenovoS3Bucket(stack, 'bucket', {
    lifecycleRules: [{
      enabled: true,
      abortIncompleteMultipartUploadAfter: cdk.Duration.days(1),
      transitions: [{
        storageClass: s3.StorageClass.INTELLIGENT_TIERING,
        transitionAfter: cdk.Duration.days(30),
      }],
    }],
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
