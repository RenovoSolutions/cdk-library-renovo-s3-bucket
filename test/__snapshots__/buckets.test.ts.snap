// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Resources": Object {
    "bucket69437533": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "AbortIncompleteMultipartUpload": Object {
                "DaysAfterInitiation": 1,
              },
              "Status": "Enabled",
              "Transitions": Array [
                Object {
                  "StorageClass": "INTELLIGENT_TIERING",
                  "TransitionInDays": 30,
                },
              ],
            },
          ],
        },
        "OwnershipControls": Object {
          "Rules": Array [
            Object {
              "ObjectOwnership": "BucketOwnerPreferred",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "bucketPolicyB2CA767C": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "bucket69437533",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "bucket69437533",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "bucket69437533",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
